### DISCUSSION TITLE
Two questions related to color settings

### DISCUSSION THREAD

**Discussion by antodc** [doublecmd/doublecmd#1955] [Suggestions]
1) The mask for 'File types' in Settings -> Colors seems to be the only one in DC with an 'Apply' button. So, to apply changes made there one has to click 'Apply' twice, one in that mask and one in the 'general' settings mask. If one forgets one of the two 'Apply' the changes are lost. Is there any particular reason for this double 'Apply' or, maybe, the layout can be simplified?

2) In 'File panels', I would like to set the 'Cursor text' color so that it does not change the color set for the file type. I mean, I have pdf files shown in red, docx in blue and 'Cursor text' is set to black. Now, when the cursor is over a pdf or a docx file they are shown in black, while I would like them to remain red and blue respectively. Is this possible?

**Comment by j2969719** [comment]
use frame cursor, with a regular cursor this is not possible because when the colors of the files are close to the color of the cursor the text will become difficult to read

**Reply by antodc** [reply]
Thank you for the suggestion. I can try to get used to it.

**Reply by antodc** [reply]
@alexx2000 The cursor in 'Files types' takes the color defined in 'File panels', but does not get the option 'Frame cursor'. This looks very bad when colors for text and cursor are the same. Maybe in 'Files types' text and cursor might have predefined unchangeble colors.

**Reply by alexx2000** [reply]
> but does not get the option 'Frame cursor'

DC use this option. But it also can be redefined in the [columns settings](https://doublecmd.github.io/doc/en/configuration.html#ConfigColumns). Check both places.

**Reply by antodc** [reply]
Not sure I understood your point, but perhaps I did not explain correctly. Look at the screenshot. Text is black and cursor is blue. When you select a line is (almost) impossible to read the text.

![Immagine](https://github.com/user-attachments/assets/c063d35d-0e35-4b07-955a-50d42c265da1)

**Reply by ags1234** [reply]
> Not sure I understood your point, but perhaps I did not explain correctly. Look at the screenshot. Text is black and cursor is blue. When you select a line is (almost) impossible to read the text.
> 
For me, the cursor and file type colors are different, see below.

![2024-11-05_121036](https://github.com/user-attachments/assets/45eb2af2-5360-4a68-9e05-bc0675f16027)

**Reply by antodc** [reply]
Do you mind if I ask you a screenshot of your settings in File panels?

**Reply by ags1234** [reply]
Sure, here it is:

![2024-11-05_143943](https://github.com/user-attachments/assets/226716cb-392f-43c1-9e07-b7e4c9085e8c)

**Reply by alexx2000** [reply]
@antodc  What operating system do you use?

**Reply by antodc** [reply]
@alexx2000

Double Commander
Version: 1.1.19 gamma
Revision: 585
Commit: 9dd31c2e9
Build date: 2024/10/06
Lazarus: 3.7.0.0
FPC: 3.2.2
Platform: x86_64-Win64-win32/win64
OS version: Windows 10 22H2 x86_64


@ags1234 Thanks a lot! That explains the difference. I have activated "Use_Frame Cursor" and I use blue for "Cursor color" and black for "Text color" and "Cursor text". That combination is fine for "File panels", but looks ugly in the "File types" mask.

**Reply by antodc** [reply]
I am really sorry to continue this conversation, but every time I go to the "File types" settings I feel really sad.

In "File panels" I have "Cursor color" set to black and I use the "Use frame cursor" option. It looks great to me when I work with DC, but the result for the "File types" settings is the following.

<img width="1135" height="668" alt="immagine" src="https://github.com/user-attachments/assets/7bec6701-23a3-4499-b184-6e3d335fe0c6" />

This page is the only one in the settings (toghether with "File panels", of course) where the cursor changes its color from the default one to the one defined in "File panels" -> "Cursor color", but -- and this is important -- it does not respect the "File panels" -> "Use frame cursor" option.

So, please, implement one of the following.

Option a) - Use the default cursor color also in this page.

Option b) - Allow the cursor in this page to respect the choice made for "Use frame cursor".

Thanks a lot for you attention!!

**Comment by Skif-off** [comment]
1. To work with file types, only three buttons are needed: Add, Delete and Apply in one row.

**Reply by antodc** [reply]
Fine, but why do you need that “Apply” button when there is already another “Apply” button just below that one? At least in Windows.

**Reply by alexx2000** [reply]
First button saves a current item, second button apply all changes and save settings into doublecmd.xml.

**Reply by antodc** [reply]
@alexx2000 Ok, but for instance in "File association" you only have "Add" and "Remove". More or less the same in "Favorite Tabs". You never have this "extra Apply" in other settings, as far as I can see. To me, it is a bit confusing, because once you click "Apply" you expect that changes are actually applied. You do not immediately realize that you have to click another "Apply" for that to happen.
A few days ago I spent 20 minutes trying to understand why the changes were not applied. I was so focused on the first part of the mask that I did not see that there was another "Apply".
Anyway, not a major issue, I understand.

**Comment by r5to** [comment]
Would be nice to have option to choose color of text under cursor on selected line, while "inverted selection" is ON. Without it there are not too many readably color options to choose.
Or maybe divide "mark color" to "mark color background" and "mark color text" to have option for selection with background color.
For example, here would be much better to have lighter color or just white, not black:
![obraz](https://github.com/user-attachments/assets/fef24dc4-5511-4d39-a17b-6ec0a0336519)

**Reply by antodc** [reply]
In my case the text color for the selected items is always equal to the background color. How did you get two colors? :-)

**Reply by r5to** [reply]
Maybe try "use inverted selection" for colors in file panels.

**Reply by antodc** [reply]
Yes, that's what I did

![Immagine](https://github.com/user-attachments/assets/36c49e23-1cb5-4e24-a27a-b2060dd3af07)

**Reply by j2969719** [reply]
i think he just uses a regular cursor with a white cursor border, black cursor text and the same blue cursor color as the mark color for some reason.

**Reply by antodc** [reply]
In my case, all selected items still have the same white color. To replicate the image @r5to showed above, the file with the black font has to be a non-selected one. Just the cursor is over it. But that would mean that our friend got a bad combination of cursor color and cursor text that has nothing to do with the selection of items.

**Reply by r5to** [reply]
> i think he just uses a regular cursor with a white cursor border, black cursor text and the same blue cursor color as the mark color for some reason.

yeah, but cursor text is white. it is changed to black on selected row because of inverted selection.
the last one is inverted cursor with border on not selected row:
![obraz](https://github.com/user-attachments/assets/91d6dd71-4a6b-4775-a73b-7f553cdc776e)

**Reply by r5to** [reply]
> In my case, all selected items still have the same white color. To replicate the image @r5to showed above, the file with the black font has to be a non-selected one. Just the cursor is over it. But that would mean that our friend got a bad combination of cursor color and cursor text that has nothing to do with the selection of items.

the case is when i want to have cursor with background, not only with border i cant control text color when cursor is on selected row. it just get inverted.

now it could be clean what is all about:
![dc colors](https://github.com/user-attachments/assets/8be87126-1ea6-47a2-ace6-0184ba0e7872)

**Reply by r5to** [reply]
intresting... this gif shows some kind of color bug, look at CloserBash file name color.
maybe its just gif compression bug:
![obraz](https://github.com/user-attachments/assets/7cb300f1-aad9-470c-b47a-e4738cdf0bb2)

**Reply by antodc** [reply]
> now it could be clean what is all about

Yes!! Now, I can see your point :-)

When "Use inverted selection" in on, the colors of the background and of the text are "inverted" for selected items and if also the cursor inverts its default text color the final result may be bad.

Just to repeat that differently for my own understanding, you are using kind of an 'inverted cursor' with dark background and light foreground. When you combine an 'inverted cursor' with 'inverted selection' you (may) get in trouble.

Having said that, I am afraid that also your suggestion to divide "mark color" to "mark color background" and "mark color text" might not work. If the cursor keeps changing its default text color when it is over a selected file, I would expect that to overwrite your "mark color text" anyway.

My own conclusion is that the best solution would be that the Cursor Text color has no effect at all on selected items. In other words, Cursor Text color should be internally disabled by DC when the cursor in on items that have been selected (and "Use inverted selection" in on). But maybe I am also missing something.

**Reply by r5to** [reply]
> Having said that, I am afraid that also your suggestion to divide "mark color" to "mark color background" and "mark color text" might not work. If the cursor keeps changing its default text color when it is over a selected file, I would expect that to overwrite your "mark color text" anyway.
>
I think it could be good, if inverted selection is OFF.

The main goal is to have option for selection with background color change, not only text. Inverted selection gives selection with background color, but is uncontrollable and not much color schemes are working good with it. All color scheme need to be dependent to this invertion to find something usable.

**Comment by ags1234** [comment]
(Two questions related to color settings)  `+ 1`    :))

Double Commander
Version: 1.2.0 alpha
Revision: 11659
Commit: defc20560
Build date: 2024/11/09
Lazarus: 3.6.0.0
FPC: 3.2.2
Platform: i386-Win32-win32/win64
OS version: Windows 11 23H2 x86_64

I activated `Layout` - `Show free spare indicator on drive label`.
But the bar is black and I'd like to change the color.

![2024-11-09 18 02 01](https://github.com/user-attachments/assets/ac2d0898-bc86-420e-ad50-22fc7e94f8eb)

Where can I find the setting for it?

![2024-11-09_174939](https://github.com/user-attachments/assets/958a9a8b-513d-4af6-823b-7923b97776f3)

**Reply by j2969719** [reply]
options>colors>category:>dive free space..

**Reply by ags1234** [reply]
@j2969719 Do you see the option you are indicating in my screenshot?

**Reply by j2969719** [reply]
erm in your screenshot you have _file panels_ selected, not **colors**

**Reply by ags1234** [reply]
Oh, yes, found it! Sorry about that, haha!

It's just that the option is well hidden, totally counterintuitive for me.
It should be in `Colors` - `File panels`, because as a user I think like this: **I want to see the free space in the file panel**, so I expect the colors config for it to also be there with the other panel colors. LOL

**Reply by j2969719** [reply]
iirc that thing should always be belove drive buttons and above tabs and filepanel

**Reply by alexx2000** [reply]
The free space indicator is not a part of the file panel. The file panel is what you see in the tab.
